crypto/params.c

```diff
@@ -0,0 +1,375 @@
+        return 0;
+    memcpy(q, p->buffer, p->buffer_size);
+    return 1;
+}
+
+int OSSL_PARAM_set_utf8_string(OSSL_PARAM *p, char *val)
+{
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, 0);
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        if (p->data_type == OSSL_PARAM_UTF8_STRING && p->buffer_size >= len) {
+            memcpy(p->buffer, val, len);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string(const char *key, void *buf,
+                                             size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_STRING, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_octet_string(const OSSL_PARAM *p, void **val, size_t max_len,
+                                size_t *used)
+{
+    size_t len;
+    char *q;
+
+    if (val == NULL || p == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+        return 0;
+
+    *used = len = p->buffer_size;
+    if (*val == NULL) {
+        if ((q = OPENSSL_malloc(len)) == NULL)
+            return 0;
+        *val = q;
+        memcpy(q, p->buffer, len);
+        return 1;
+    }
+    if (max_len < len)
+        return 0;
+    memcpy(*val, p->buffer, len);
+    return 1;
+}
+
+int OSSL_PARAM_set_octet_string(OSSL_PARAM *p, void *val, size_t len)
+{
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, 0);
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        if (p->data_type == OSSL_PARAM_OCTET_STRING && p->buffer_size >= len) {
+            memcpy(p->buffer, val, len);
+            return 1;
+        }
+    }
+    return 0;
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string_ptr(const char *key, char **buf,
+                                                size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UTF8_PTR, buf, bsize,
+                                rsize);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_octet_string_ptr(const char *key, void **buf,
+                                                 size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_OCTET_PTR, buf, bsize,
+                                rsize);
+}
+
+int OSSL_PARAM_get_utf8_ptr(const OSSL_PARAM *p, char **val)
+{
+    if (val == NULL || p == NULL || p->data_type != OSSL_PARAM_UTF8_PTR)
+        return 0;
+    *val = *(char **)p->buffer;
+    return 1;
+}
+
+int OSSL_PARAM_set_utf8_ptr(OSSL_PARAM *p, char *val)
+{
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, 0);
+    if (val != NULL) {
+        const size_t len = strlen(val) + 1;
+
+        SET_RETURN_SIZE(p, len);
+        if (p->data_type == OSSL_PARAM_UTF8_PTR) {
+            *(char **)p->buffer = val;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+int OSSL_PARAM_get_octet_ptr(const OSSL_PARAM *p, void **val)
```

Comment:
Shouldn't you also have a `size_t *` parameter to report back the length?

Reply:
Yes, added.